File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/api/graphql/datafetcher/PostRepresentationFetcher.java
package com.github.arhor.aws.graphql.federation.comments.api.graphql.datafetcher;

import com.github.arhor.aws.graphql.federation.comments.api.graphql.dataloader.PostRepresentationBatchLoader;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.DgsConstants.POST;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.Post;
import com.github.arhor.aws.graphql.federation.comments.service.PostRepresentationService;
import com.github.arhor.aws.graphql.federation.starter.security.CurrentUserDetails;
import com.github.arhor.aws.graphql.federation.starter.tracing.Trace;
import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsDataFetchingEnvironment;
import com.netflix.graphql.dgs.DgsEntityFetcher;
import com.netflix.graphql.dgs.DgsMutation;
import com.netflix.graphql.dgs.InputArgument;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

import static com.github.arhor.aws.graphql.federation.common.MapExtKt.getUuid;

@Trace
@DgsComponent
@RequiredArgsConstructor
public class PostRepresentationFetcher {

    private final PostRepresentationService postService;

    /* ---------- Entity Fetchers ---------- */

    @DgsEntityFetcher(name = POST.TYPE_NAME)
    public CompletableFuture<Post> resolvePost(final Map<String, ?> values, final DgsDataFetchingEnvironment dfe) {
        final var postId = getUuid(values, POST.Id);
        final var loader = dfe.<UUID, Post>getDataLoader(PostRepresentationBatchLoader.class);

        return loader.load(postId);
    }

    /* ---------- Mutations ---------- */

    @DgsMutation
    @PreAuthorize("isAuthenticated()")
    public boolean togglePostComments(
        final @InputArgument UUID postId,
        final @AuthenticationPrincipal CurrentUserDetails actor
    ) {
        return postService.togglePostComments(postId, actor);
    }
}

File: app-service-votes/src/main/kotlin/com/github/arhor/aws/graphql/federation/votes/api/graphql/datafetcher/PostRepresentationFetcher.kt
package com.github.arhor.aws.graphql.federation.votes.api.graphql.datafetcher

import com.github.arhor.aws.graphql.federation.common.getUuid
import com.github.arhor.aws.graphql.federation.votes.api.graphql.dataloader.PostRepresentationBatchLoader
import com.github.arhor.aws.graphql.federation.votes.generated.graphql.DgsConstants.POST
import com.github.arhor.aws.graphql.federation.votes.generated.graphql.types.Post
import com.netflix.graphql.dgs.DgsComponent
import com.netflix.graphql.dgs.DgsDataFetchingEnvironment
import com.netflix.graphql.dgs.DgsEntityFetcher
import java.util.UUID
import java.util.concurrent.CompletableFuture

@DgsComponent
class PostRepresentationFetcher {

    /* ---------- Entity Fetchers ---------- */

    @DgsEntityFetcher(name = POST.TYPE_NAME)
    fun resolvePost(values: Map<String, Any>, dfe: DgsDataFetchingEnvironment): CompletableFuture<Post> {
        val postId = values.getUuid(POST.Id)
        val loader = dfe.getDataLoader<UUID, Post>(PostRepresentationBatchLoader::class.java)

        return loader.load(postId)
    }
}
